File: python/dolfinx/multigrid.py
# Copyright (C) 2025 Paul T. Kühner
#
# This file is part of DOLFINx (https://www.fenicsproject.org)
#
# SPDX-License-Identifier:    LGPL-3.0-or-later

import numpy as np
from numpy.typing import NDArray

from dolfinx.cpp.multigrid import inclusion_mapping_float32 as _inclusion_mapping_float32
from dolfinx.cpp.multigrid import inclusion_mapping_float64 as _inclusion_mapping_float64
from dolfinx.mesh import Mesh

__all__ = ["inclusion_mapping"]


def inclusion_mapping(mesh_from: Mesh, mesh_to: Mesh) -> NDArray[np.int64]:
    if np.issubdtype(mesh_from.geometry.x.dtype, np.float32):
        return _inclusion_mapping_float32(mesh_from._cpp_object, mesh_to._cpp_object)
    elif np.issubdtype(mesh_from.geometry.x.dtype, np.float64):
        return _inclusion_mapping_float64(mesh_from._cpp_object, mesh_to._cpp_object)
    else:
        raise RuntimeError("Unsupported mesh dtype.")

File: python/dolfinx/wrappers/multigrid.cpp
// Copyright (C) 2025 Paul T. Kühner
//
// This file is part of DOLFINX (https://www.fenicsproject.org)
//
// SPDX-License-Identifier:    LGPL-3.0-or-later

#include <concepts>

#include <nanobind/nanobind.h>
#include <nanobind/stl/vector.h>

#include <dolfinx/la/MatrixCSR.h>
#include <dolfinx/multigrid/inclusion.h>

#include "dolfinx_wrappers/array.h"

namespace nb = nanobind;

namespace dolfinx_wrappers
{

template <std::floating_point T>
void declare_inlcusion_mapping(nb::module_& m, const std::string& type)
{
  
  m.def(
      ("inclusion_mapping_" + type).c_str(),
      [](const dolfinx::mesh::Mesh<T>& mesh_from,
         const dolfinx::mesh::Mesh<T>& mesh_to)
      {
        return dolfinx_wrappers::as_nbarray(
            dolfinx::multigrid::inclusion_mapping<T>(mesh_from, mesh_to));
      },
      nb::arg("mesh_from"), nb::arg("mesh_to"),
      "Computes inclusion mapping between two meshes");
}

void multigrid(nb::module_& m)
{
  declare_inlcusion_mapping<float>(m, "float32");
  declare_inlcusion_mapping<double>(m, "float64");
}

} // namespace dolfinx_wrappers
